File: CS/WindowsApplication3/FilterNameProvider.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using DevExpress.Utils.Menu;
using DevExpress.Utils.Serializing;
using DevExpress.XtraBars.Docking2010.Customization;
using DevExpress.XtraBars.Docking2010.Views.WindowsUI;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;

namespace DXSample {
    public class FilterNameProvider {
        List<GridFitlerItem> filters;
        DXPopupMenu menu;
        GridView view;
        public FilterNameProvider(GridView view) {
            this.view = view;
            view.OptionsLayout.StoreDataSettings = true;
            filters = new List<GridFitlerItem>();
        }
        bool fAllowSettingFilterNames;
        public bool AllowSettingFilterNames {
            get {
                return fAllowSettingFilterNames;
            }
            set {
                if (fAllowSettingFilterNames != value) {
                    fAllowSettingFilterNames = value;
                    if (fAllowSettingFilterNames) {
                        SubscribeToEvents();
                        CreatePopupMenu();
                    }
                    else {
                        UnsubscribeFromEvents();
                        DestroyPopupMenu();
                    }
                }
            }
        }
        [XtraSerializableProperty(XtraSerializationVisibility.Collection, true)]
        public List<GridFitlerItem> GridFilters {
            get { return filters; }
        }
        internal GridFitlerItem XtraCreateGridFiltersItem(XtraItemEventArgs e) {
            GridFitlerItem fItem = new GridFitlerItem();
            GridFilters.Add(fItem);
            return fItem;
        }
        void SubscribeToEvents() {
            view.MouseDown += OnViewMouseDown;
            view.CustomFilterDisplayText += OnViewCustomFilterDisplayText;
        }
        void OnViewCustomFilterDisplayText(object sender, DevExpress.XtraEditors.Controls.ConvertEditValueEventArgs e) {
            GridView view = sender as GridView;
            string filterString = e.Value.ToString();
            GridFitlerItem fItem = filters.FirstOrDefault(it => it.FilterString == filterString);
            if(fItem != null){
                e.Value = fItem.FilterName;
                e.Handled = true;
            }
        }
        void OnViewMouseDown(object sender, System.Windows.Forms.MouseEventArgs e) {
            GridView view = sender as GridView;
            GridHitInfo hitInfo = view.CalcHitInfo(e.Location);
            if (e.Button == MouseButtons.Right && hitInfo.HitTest == GridHitTest.FilterPanelText) {
                ((IDXDropDownControl)menu).Show(new SkinMenuManager(view.GridControl.LookAndFeel), view.GridControl, e.Location);
            }
        }
        void CreatePopupMenu() {
            menu = new DXPopupMenu();
            menu.Items.Add(new DXMenuItem("Save Filter As", OnSaveItemClick));
        }
        void OnSaveItemClick(object sender, EventArgs e) {
            FlyoutAction action = new FlyoutAction() { Caption = "Save Filter As", Description = "" };
            FlyoutCommand command1 = new FlyoutCommand() { Text = "Save", Result = DialogResult.OK };
            FlyoutCommand command2 = new FlyoutCommand() { Text = "Cancel", Result = DialogResult.Cancel };
            action.Commands.Add(command1);
            action.Commands.Add(command2);
            using (TextEdit filterNameEdit = new TextEdit()) {
                filterNameEdit.Width = 300;
                if (FlyoutDialog.Show(view.GridControl.FindForm(), action, filterNameEdit) == System.Windows.Forms.DialogResult.OK) {
                    if (string.IsNullOrEmpty(filterNameEdit.Text.Trim())) return;
                    GridFitlerItem fItem = filters.FirstOrDefault(it => it.FilterString == view.ActiveFilterString);
                    if (fItem == null)
                        filters.Add(new GridFitlerItem() { FilterString = view.ActiveFilterString, FilterName = filterNameEdit.Text });
                    else
                        fItem.FilterName = filterNameEdit.Text;
                }
            }
        }
        void UnsubscribeFromEvents() {
            view.MouseDown -= OnViewMouseDown;
            view.CustomFilterDisplayText -= OnViewCustomFilterDisplayText;
        }
        void DestroyPopupMenu() {
            menu.Dispose();
            menu = null;
        }
    }
    public class GridFitlerItem {
        public GridFitlerItem() { }
        [XtraSerializableProperty]
        public string FilterString { get; set; }
        [XtraSerializableProperty]
        public string FilterName { get; set; }
    }
    public static class Extensions {
        public static void SaveLayoutToXmlEx(this ColumnView view, FilterNameProvider provider, string filePath) {
            XmlXtraSerializer serializer = new XmlXtraSerializer();
            serializer.SerializeObjects(new XtraObjectInfo[] {new XtraObjectInfo("View", view), new XtraObjectInfo("Filters", provider) }, filePath, "DXSample");
        }
        public static void RestoreLayoutFromXmlEx(this ColumnView view, FilterNameProvider provider, string filePath) {
            XmlXtraSerializer serializer = new XmlXtraSerializer();
            serializer.DeserializeObjects(new XtraObjectInfo[] { new XtraObjectInfo("View", view), new XtraObjectInfo("Filters", provider) }, filePath, "DXSample");
        }
    }
}

File: VB/WindowsApplication3/FilterNameProvider.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.Utils.Menu
Imports DevExpress.Utils.Serializing
Imports DevExpress.XtraBars.Docking2010.Customization
Imports DevExpress.XtraBars.Docking2010.Views.WindowsUI
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Namespace DXSample
	Public Class FilterNameProvider
		Private filters As List(Of GridFitlerItem)
		Private menu As DXPopupMenu
		Private view As GridView
		Public Sub New(ByVal view As GridView)
			Me.view = view
			view.OptionsLayout.StoreDataSettings = True
			filters = New List(Of GridFitlerItem)()
		End Sub
		Private fAllowSettingFilterNames As Boolean
		Public Property AllowSettingFilterNames() As Boolean
			Get
				Return fAllowSettingFilterNames
			End Get
			Set(ByVal value As Boolean)
				If fAllowSettingFilterNames <> value Then
					fAllowSettingFilterNames = value
					If fAllowSettingFilterNames Then
						SubscribeToEvents()
						CreatePopupMenu()
					Else
						UnsubscribeFromEvents()
						DestroyPopupMenu()
					End If
				End If
			End Set
		End Property
		<XtraSerializableProperty(XtraSerializationVisibility.Collection, True)>
		Public ReadOnly Property GridFilters() As List(Of GridFitlerItem)
			Get
				Return filters
			End Get
		End Property
		Friend Function XtraCreateGridFiltersItem(ByVal e As XtraItemEventArgs) As GridFitlerItem
			Dim fItem As New GridFitlerItem()
			GridFilters.Add(fItem)
			Return fItem
		End Function
		Private Sub SubscribeToEvents()
			AddHandler view.MouseDown, AddressOf OnViewMouseDown
			AddHandler view.CustomFilterDisplayText, AddressOf OnViewCustomFilterDisplayText
		End Sub
		Private Sub OnViewCustomFilterDisplayText(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.ConvertEditValueEventArgs)
			Dim view As GridView = TryCast(sender, GridView)
			Dim filterString As String = e.Value.ToString()
			Dim fItem As GridFitlerItem = filters.FirstOrDefault(Function(it) it.FilterString = filterString)
			If fItem IsNot Nothing Then
				e.Value = fItem.FilterName
				e.Handled = True
			End If
		End Sub
		Private Sub OnViewMouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
			Dim view As GridView = TryCast(sender, GridView)
			Dim hitInfo As GridHitInfo = view.CalcHitInfo(e.Location)
			If e.Button = MouseButtons.Right AndAlso hitInfo.HitTest = GridHitTest.FilterPanelText Then
				CType(menu, IDXDropDownControl).Show(New SkinMenuManager(view.GridControl.LookAndFeel), view.GridControl, e.Location)
			End If
		End Sub
		Private Sub CreatePopupMenu()
			menu = New DXPopupMenu()
			menu.Items.Add(New DXMenuItem("Save Filter As", AddressOf OnSaveItemClick))
		End Sub
		Private Sub OnSaveItemClick(ByVal sender As Object, ByVal e As EventArgs)
			Dim action As New FlyoutAction() With {
				.Caption = "Save Filter As",
				.Description = ""
			}
			Dim command1 As New FlyoutCommand() With {
				.Text = "Save",
				.Result = DialogResult.OK
			}
			Dim command2 As New FlyoutCommand() With {
				.Text = "Cancel",
				.Result = DialogResult.Cancel
			}
			action.Commands.Add(command1)
			action.Commands.Add(command2)
			Using filterNameEdit As New TextEdit()
				filterNameEdit.Width = 300
				If FlyoutDialog.Show(view.GridControl.FindForm(), action, filterNameEdit) = System.Windows.Forms.DialogResult.OK Then
					If String.IsNullOrEmpty(filterNameEdit.Text.Trim()) Then
						Return
					End If
					Dim fItem As GridFitlerItem = filters.FirstOrDefault(Function(it) it.FilterString = view.ActiveFilterString)
					If fItem Is Nothing Then
						filters.Add(New GridFitlerItem() With {
							.FilterString = view.ActiveFilterString,
							.FilterName = filterNameEdit.Text
						})
					Else
						fItem.FilterName = filterNameEdit.Text
					End If
					view.LayoutChanged()
				End If
			End Using
		End Sub
		Private Sub UnsubscribeFromEvents()
			RemoveHandler view.MouseDown, AddressOf OnViewMouseDown
			RemoveHandler view.CustomFilterDisplayText, AddressOf OnViewCustomFilterDisplayText
		End Sub
		Private Sub DestroyPopupMenu()
			menu.Dispose()
			menu = Nothing
		End Sub
	End Class
	Public Class GridFitlerItem
		Public Sub New()
		End Sub
		<XtraSerializableProperty>
		Public Property FilterString() As String
		<XtraSerializableProperty>
		Public Property FilterName() As String
	End Class
	Public Module Extensions
		<System.Runtime.CompilerServices.Extension> _
		Public Sub SaveLayoutToXmlEx(ByVal view As ColumnView, ByVal provider As FilterNameProvider, ByVal filePath As String)
			Dim serializer As New XmlXtraSerializer()
			serializer.SerializeObjects(New XtraObjectInfo() {
				New XtraObjectInfo("View", view),
				New XtraObjectInfo("Filters", provider)
			}, filePath, "DXSample")
		End Sub
		<System.Runtime.CompilerServices.Extension> _
		Public Sub RestoreLayoutFromXmlEx(ByVal view As ColumnView, ByVal provider As FilterNameProvider, ByVal filePath As String)
			Dim serializer As New XmlXtraSerializer()
			serializer.DeserializeObjects(New XtraObjectInfo() {
				New XtraObjectInfo("View", view),
				New XtraObjectInfo("Filters", provider)
			}, filePath, "DXSample")
		End Sub
	End Module
End Namespace
